### PR TITLE
Report error on method name for chained method calls

### LINKED ISSUE #31365: Improve error messages for method chains
## Search Terms

error call chain fluent api

## Suggestion

In call chains some errors like TS2554 ("Expected x arguments, but got y") are reported for the whole call chain.

Example:

image

In this case the error is reported (starting) at line 5 where the actual error is in line 14.

*Suggestion:* Just highlight the concerned (last) method invocation in this case.  

## Use Cases

* Fluent APIs: These sub-optimal error reporting is misguiding in fluent APIs because the start line number can be off hundreds of lines. Further, way too much is highlighted in red squiggles (for a single missing argument as in the example above).

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by dhruvrajvanshi**
Fixes #31365

This is a first stab at fixing the issue. It reports the error at the name of the final method call.

Ideally, the error would be reported at the argument list span for the call but I couldn't find a straightforward way to get a span for the argument list. I'm willing to the PR if we really need reporting at the argument list.

### CODE REVIEW

File: tests/baselines/reference/callWithMissingVoid.errors.txt [hunk 1]
```diff
@@ -28,19 +28,19 @@ tests/cases/conformance/expressions/functionCalls/callWithMissingVoid.ts(75,1):
     
     declare const xAny: X<any>;
     xAny.f() // error, any still expects an argument
-    ~~~~~~~~
+         ~
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/callWithMissingVoid.errors.txt]: I think the idea was that the argument list was supposed to be included in the range.

Comment (DanielRosenwasser) [reply tests/baselines/reference/callWithMissingVoid.errors.txt]: Ah, just saw your comment in the PR.

**Comment by DanielRosenwasser** [comment]
Spoke with @RyanCavanaugh - we think just erroring on the name itself is fine, but we'd like to make this consistent for plain function calls as well.

**Comment by dhruvrajvanshi** [comment]
@DanielRosenwasser I'm not quite sure I understand.
Do you mean something like this?

```ts
function getCallErrorNode(node: CallLikeExpression): Node {
        if (isCallExpression(node)) {
            if (isPropertyAccessExpression(node.expression)) {
                return node.expression.name;
            } else {
                return node.expression
            }
        }
                
        return node;
}
```

Because this is breaking a lot of error messages (and they look worse IMO).

Ideally, the error should include the final method name and its argument list which could be done by including a span in `NodeArray`.

**Comment by DanielRosenwasser** [comment]
Yeah, something like that code.

> Because this is breaking a lot of error messages (and they look worse IMO).

Maybe we'd do this for only `Identifier` nodes. What looks grosser?

**Comment by dhruvrajvanshi** [comment]
I've updated the PR so you can see for yourself. Most call argument arity errors exclude the argument list. It is more uniform with the new change though.

Is there a way to report diagnostics on spans instead of nodes?

**Comment by DanielRosenwasser** [comment]
I think this is a great first step. If we want to iterate further, we'll take another PR.

**Pull request merged**